### DISCUSSION TITLE
Any way to modify timestamps of certain groups without affecting data sampling or group/channel meta data

### DISCUSSION THREAD

**Discussion by motthomasn** [danielhrisca/asammdf#1010] [Q&A] [answered]
Hi all, I'm looking to adjust the timestamps of all the channel groups in an mf4 file without affecting the sampling or the file meta data. Is this something that can currently be done with asammdf? By adjust, I specifically mean offset. 

What I am ultimately trying to achieve is to time-align and merge 2 mf4 files captured from 2 different sources. I am currently loading the alignment signal from both files and am able to determine the time shift needed for best alignment. I could just load the 2nd file data as a dataframe, shift the timestamps and save the dataframe as a new mf4 file but then I would lose all the metadata from the channels. What I'd rather do is somehow modify the timestamps in isolation with the least amount of disturbance to any other data in the file and then merge the 2 files which I am comfortable in doing.

I'm aware that there is a similar question already asked a few days ago ([https://github.com/danielhrisca/asammdf/discussions/1009](url)) but I feel like my query is different enough to not derail their question.

**Comment by danielhrisca** [comment]
Please show how the time channel conversion rule is stored in the file. For example

![image](https://github.com/danielhrisca/asammdf/assets/20952040/218e6d3e-0bcb-4730-9769-98ab494d0b28)
 or by double clicking in the asammdf gui

![image](https://github.com/danielhrisca/asammdf/assets/20952040/dc11d132-6522-4001-a181-a8ec76789702)

**Reply by motthomasn** [reply]
Hi Daniel. There is no conversion stored in the file so I'm assuming it is stored as a float type.
![image](https://github.com/danielhrisca/asammdf/assets/65952421/b019fdd4-472b-4d83-94bc-8d4e6c65114e)
![image](https://github.com/danielhrisca/asammdf/assets/65952421/5305c76c-2659-417b-b2e1-5071baea8c2c)

**Reply by danielhrisca** [reply, marked as answer]
In this case you have two options

## !!! try them on copies of the original file


1. load the file and set a linear conversion for the time channel, then save the file again. This will change the internal layout of the file , but it will preserve the metadata 

```python

from asammdf import MDF
from asammdf.blocks.conversion_utils import from_dict
mdf = MDF(file_name)
time = mdf.groups[0].channels[0]
time.conversion = from_dict({'a': 1, 'b': offset})  # linear conversion y = a *x + b
mdf.save(file_name, overwrite=True, compression=2)
mdf.close()

```

2. lower level get your hands dirty and create a new conversion block that will be appended to the end of the file, then link it to the time channel 

```python
from asammdf import MDF
from asammdf.blocks.conversion_utils import from_dict
from struct import pack

mdf = MDF(file_name)
time = mdf.groups[0].channels[0]
time_conv_link = time.address + 56
mdf.close()

conversion =  from_dict({'a': 1, 'b': offset}) 
with open(file_name, 'rb+') as f:
    f.seek(0, 2)
    addr = f.tell()
    if addr % 8:
        f.write(bytes(8 - (addr % 8)))
    conv_addr = f.tell()
    f.write(bytes(conversion))
    f.seek(time_conv_link)
    f.write(pack('<Q', conv_addr))
   
```

**Reply by motthomasn** [reply]
Brilliant! Thanks Daniel. Both options work a treat if I keep the 2 files seperate and display them together. They do seems to fall over when I try to stack the 2 files into one for convenient storage so thats my next challenge to figure out. The files stack OK but the timestamps that were offset are then extremely large so the data is no longer aligned after stacking. There is something that I am not understanding about how the timestamps are defined in the file I think.

Can ask, regarding option 2, why do you offset the time.address by 54 to get time_conv_link? Is this a required offset or arbitrary?

**Reply by danielhrisca** [reply]
when you open a file all the timestamp are relative to the measurement start `mdf.header.start_time`.

Did you already try to just stack the initial files directly? this already uses the start_time info

**Reply by motthomasn** [reply]
Ah OK. Then modifying the start_time info on the 2nd file is actually the easiest route in the end. Thank you very much!

What I've ended up with is:
```python3
from asammdf import MDF
from datetime import timedelta

mdf1 = MDF(file1path)
mdf2 = MDF(file2path)

mdf2.start_time = mdf1.start_time + timedelta(seconds=timeOffset)

mdf3 = MDF.stack([mdf1, mdf2])
mdf3.save("test_offset_merge.mf4", overwrite=True, compression=2)
mdf3.close()
```

I have in the past needed to time offset individual channels though so your original asnswer is also very useful.
